File: src/palimpsest/text/segment.py
"""Segmentation of an essay into paragraphs and sentences, preserving character offsets.

Every downstream layer -- feature extraction, scoring, and the highlighting in the web UI --
addresses text by ``(start, end)`` offsets into the *original* string. So the one invariant
this module must never break is:

    text[span.start:span.end] == span.text

We deliberately do not pull in spaCy or NLTK. Sentence splitting for essay prose is a small,
well-understood problem, and a 400 MB dependency whose model downloads at import time would
be the single largest thing in the project for the least benefit. The rules below are written
out explicitly so they can be argued with -- see ``tests/test_segment.py``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

__all__ = ["Span", "split_sentences", "split_paragraphs", "tokenize_words"]


@dataclass(frozen=True, slots=True)
class Span:
    """A half-open character range ``[start, end)`` into a source string."""

    start: int
    end: int
    text: str

    def __len__(self) -> int:
        return self.end - self.start


# Tokens that end in '.' but almost never end a sentence. Kept lowercase; matched case-folded.
_ABBREVIATIONS = frozenset(
    """
    mr mrs ms dr prof sr jr st rev hon gen col capt lt sgt gov pres supt
    inc ltd co corp dept univ assn bros
    jan feb mar apr jun jul aug sep sept oct nov dec
    mon tue tues wed thu thurs fri sat sun
    vs etc al ca cf ed eds vol no pp fig figs approx
    """.split()
)

# Multi-dot abbreviations such as "e.g." / "U.S." / "Ph.D." / "a.m.".
_DOTTED_ABBREV = re.compile(r"(?:[A-Za-z]\.){2,}$")

# A sentence terminator, optionally followed by closing quotes/brackets.
_TERMINATOR = re.compile(r"[.!?…]+[\"'”’)\]]*")

# Paragraphs are separated by a blank line, or by a single newline when the text uses
# one-line-per-paragraph formatting (common when an essay is pasted out of a text box).
_PARA_SPLIT = re.compile(r"\n\s*\n")

_WORD = re.compile(r"[A-Za-zÀ-ɏ']+")


def split_paragraphs(text: str) -> list[Span]:
    """Split ``text`` on blank lines into non-empty paragraph spans."""
    spans: list[Span] = []
    cursor = 0
    for chunk in _PARA_SPLIT.split(text):
        start = text.index(chunk, cursor) if chunk else cursor
        stripped = chunk.strip()
        if stripped:
            offset = chunk.index(stripped)
            spans.append(Span(start + offset, start + offset + len(stripped), stripped))
        cursor = start + len(chunk)
    return spans


def _is_abbreviation(text: str, dot_index: int) -> bool:
    """True if the '.' at ``dot_index`` closes an abbreviation rather than a sentence."""
    # Walk back over the word that the dot terminates.
    i = dot_index
    while i > 0 and (text[i - 1].isalpha() or text[i - 1] == "."):
        i -= 1
    word = text[i : dot_index + 1]
    if _DOTTED_ABBREV.match(word):
        return True
    return word.rstrip(".").lower() in _ABBREVIATIONS


def _is_decimal(text: str, dot_index: int) -> bool:
    """True for the '.' in '3.14' -- digits on both sides."""
    return (
        dot_index > 0
        and dot_index + 1 < len(text)
        and text[dot_index - 1].isdigit()
        and text[dot_index + 1].isdigit()
    )


def _is_initial(text: str, dot_index: int) -> bool:
    """True for the '.' after a single capital letter, as in 'J. K. Rowling'."""
    if dot_index == 0 or not text[dot_index - 1].isupper():
        return False
    before = dot_index - 2
    return before < 0 or not text[before].isalpha()


def _boundary_is_real(text: str, end: int) -> bool:
    """True if a sentence genuinely ends at offset ``end`` (exclusive)."""
    if end >= len(text):
        return True
    # Require whitespace after the terminator; "google.com" must not split.
    if not text[end].isspace():
        return False
    # The next non-space character should start a new sentence. Lowercase usually means
    # we cut something we should not have (e.g. an abbreviation we do not know about),
    # but a newline is strong enough evidence on its own.
    rest = text[end:]
    if "\n" in rest[: len(rest) - len(rest.lstrip())]:
        return True
    nxt = rest.lstrip()
    if not nxt:
        return True
    return not (nxt[0].isalpha() and nxt[0].islower())


def split_sentences(text: str) -> list[Span]:
    """Split ``text`` into sentence spans.

    Sentences never cross a paragraph boundary: a paragraph break is always a sentence
    break, even if the writer omitted the full stop. Leading/trailing whitespace is
    excluded from every span so that ``span.text`` is exactly the sentence.
    """
    spans: list[Span] = []
    for para in split_paragraphs(text):
        spans.extend(_split_paragraph_sentences(para))
    return spans


def _split_paragraph_sentences(para: Span) -> list[Span]:
    body = para.text
    out: list[Span] = []
    start = 0

    for match in _TERMINATOR.finditer(body):
        dot = match.start()
        end = match.end()
        char = body[dot]
        if char == "." and (
            _is_decimal(body, dot) or _is_initial(body, dot) or _is_abbreviation(body, dot)
        ):
            continue
        if not _boundary_is_real(body, end):
            continue
        piece = body[start:end]
        stripped = piece.strip()
        if stripped:
            lead = piece.index(stripped[0]) if stripped else 0
            out.append(
                Span(para.start + start + lead, para.start + start + lead + len(stripped), stripped)
            )
        start = end

    tail = body[start:]
    stripped = tail.strip()
    if stripped:
        lead = tail.index(stripped[0])
        out.append(
            Span(para.start + start + lead, para.start + start + lead + len(stripped), stripped)
        )
    return out


def tokenize_words(text: str) -> list[str]:
    """Lowercased alphabetic word tokens. Used by the n-gram and lexicon features."""
    return [m.group(0).lower() for m in _WORD.finditer(text)]

File: edge/src/segment.js
/**
 * Port of `palimpsest/text/segment.py`.
 *
 * The invariant the Python module states is the one that matters here too:
 *
 *     text.slice(span.start, span.end) === span.text
 *
 * One deliberate difference from Python, recorded rather than hidden. Python's `len()` and
 * slicing count Unicode code points; JavaScript counts UTF-16 code units. They agree on
 * everything in the Basic Multilingual Plane and disagree on astral characters — emoji,
 * rarer CJK. This port is UTF-16 throughout, which makes it *self*-consistent with the
 * observer (whose offsets have always been produced by `String.indexOf` inside the Worker)
 * and with the browser that slices the text to draw highlights.
 *
 * The Python build mixes the two conventions at that seam, so an essay containing an emoji
 * would shift its highlights there. No document in the parity corpus contains one, so this
 * costs nothing measurable; see edge/PARITY.md.
 */

import {
  isAlpha, isDigit, isLower, isSpace, isUpper, leadingSpace, PY_SPACE_CLASS, rstripDots,
} from './pyshim.js';

const ABBREVIATIONS = new Set(
  `mr mrs ms dr prof sr jr st rev hon gen col capt lt sgt gov pres supt
   inc ltd co corp dept univ assn bros
   jan feb mar apr jun jul aug sep sept oct nov dec
   mon tue tues wed thu thurs fri sat sun
   vs etc al ca cf ed eds vol no pp fig figs approx`.split(/\s+/).filter(Boolean),
);

const DOTTED_ABBREV = /^(?:[A-Za-z]\.){2,}$/;
const TERMINATOR = /[.!?…]+["'”’)\]]*/g;
// Python's `\n\s*\n`, with `\s` spelled out: JavaScript's `\s` would both miss U+0085 and
// match U+FEFF, and this regex decides where every paragraph starts.
const PARA_SPLIT = new RegExp(`\\n${PY_SPACE_CLASS}*\\n`);
const WORD = /[A-Za-zÀ-ɏ']+/g;

export function splitParagraphs(text) {
  const spans = [];
  let cursor = 0;
  for (const chunk of text.split(PARA_SPLIT)) {
    const start = chunk ? text.indexOf(chunk, cursor) : cursor;
    const lead = leadingSpace(chunk);
    let end = chunk.length;
    while (end > lead && isSpace(chunk[end - 1])) end -= 1;
    const stripped = chunk.slice(lead, end);
    if (stripped) {
      spans.push({ start: start + lead, end: start + lead + stripped.length, text: stripped });
    }
    cursor = start + chunk.length;
  }
  return spans;
}

function isAbbreviation(text, dotIndex) {
  let i = dotIndex;
  while (i > 0 && (isAlpha(text[i - 1]) || text[i - 1] === '.')) i -= 1;
  const word = text.slice(i, dotIndex + 1);
  if (DOTTED_ABBREV.test(word)) return true;
  return ABBREVIATIONS.has(rstripDots(word).toLowerCase());
}

const isDecimal = (text, i) =>
  i > 0 && i + 1 < text.length && isDigit(text[i - 1]) && isDigit(text[i + 1]);

function isInitial(text, i) {
  if (i === 0 || !isUpper(text[i - 1])) return false;
  const before = i - 2;
  return before < 0 || !isAlpha(text[before]);
}

function boundaryIsReal(text, end) {
  if (end >= text.length) return true;
  if (!isSpace(text[end])) return false;
  const rest = text.slice(end);
  const lead = leadingSpace(rest);
  if (rest.slice(0, lead).includes('\n')) return true;
  const next = rest.slice(lead);
  if (!next) return true;
  return !(isAlpha(next[0]) && isLower(next[0]));
}

function splitParagraphSentences(para) {
  const body = para.text;
  const out = [];
  let start = 0;

  TERMINATOR.lastIndex = 0;
  let match;
  while ((match = TERMINATOR.exec(body)) !== null) {
    const dot = match.index;
    const end = match.index + match[0].length;
    if (match[0].length === 0) {
      TERMINATOR.lastIndex += 1;
      continue;
    }
    if (
      body[dot] === '.' &&
      (isDecimal(body, dot) || isInitial(body, dot) || isAbbreviation(body, dot))
    ) {
      continue;
    }
    if (!boundaryIsReal(body, end)) continue;

    const piece = body.slice(start, end);
    const lead = leadingSpace(piece);
    let tail = piece.length;
    while (tail > lead && isSpace(piece[tail - 1])) tail -= 1;
    const stripped = piece.slice(lead, tail);
    if (stripped) {
      out.push({
        start: para.start + start + lead,
        end: para.start + start + lead + stripped.length,
        text: stripped,
      });
    }
    start = end;
  }

  const tailPiece = body.slice(start);
  const lead = leadingSpace(tailPiece);
  let tailEnd = tailPiece.length;
  while (tailEnd > lead && isSpace(tailPiece[tailEnd - 1])) tailEnd -= 1;
  const stripped = tailPiece.slice(lead, tailEnd);
  if (stripped) {
    out.push({
      start: para.start + start + lead,
      end: para.start + start + lead + stripped.length,
      text: stripped,
    });
  }
  return out;
}

export function splitSentences(text) {
  const spans = [];
  for (const para of splitParagraphs(text)) spans.push(...splitParagraphSentences(para));
  return spans;
}

/** Lowercased alphabetic word tokens, as the n-gram and lexicon features consume them. */
export function tokenizeWords(text) {
  WORD.lastIndex = 0;
  const out = [];
  let m;
  while ((m = WORD.exec(text)) !== null) out.push(m[0].toLowerCase());
  return out;
}
